### PR TITLE
Creating SAM Handler

### PR DESCRIPTION
**Pull request by jfalkenstein**
## In this PR:
* Introducing the Sceptre SAM Handler!

### CODE REVIEW

File: setup.py
Comment (jfalkenstein) [inline setup.py:21]: Does this really need to be on the setup? I don't particularly want to put my email address on github to be scraped by all the world's bots...

File: requirements.txt
Comment (jfalkenstein) [inline requirements.txt]: Since this is invoked with subprocess, it doesn't make sense to pin this.

File: README.md
Comment (zaro0508) [inline README.md]: would be nice to match with repo name or vise versa.  should we change the repo name to match or change this to match repo name?

File: README.md
Comment (zaro0508) [inline README.md:98]: We have a global lambda repo (S3 bucket) that we deploy our lambdas into which is not in the same account as the one where we want to deploy our lambdas into.   Can we pass in a different profile and region so it gets deployed to a separate account?

File: requirements.txt
Comment (zaro0508) [reply requirements.txt]: hmm, aws-sam-cli from pypi does seem to work however the official sam cli docs does not recommend installing from pypi.  I'm wondering if that could cause problems later?

File: setup.py
Comment (zaro0508) [reply setup.py:21]: i don't think it's required but also don't see why you can't just leave as `sceptre@sceptre.org`?

File: setup.py
Comment (zaro0508) [inline setup.py]: curious why you shorted it from `TEMPLATE_HANDLER_TYPE` to `TEMPLATE_TYPE`?

File: setup.py
Comment (zaro0508) [inline setup.py]: maybe add aws and cloudformation?

File: README.md
Comment (zaro0508) [inline README.md:71]: In order to invoke `sam build` users would need to have already setup all the library dependencies that the app requires correct?  That seems to imply that there needs to be some instructions to bootstrap an  environment or vritual env for doing the build and package?

File: README.md
Comment (zaro0508) [inline README.md]: it's already a sam handler so maybe just name this `path:`?

File: README.md
Comment (zaro0508) [inline README.md]: same here, `build_args`?

File: README.md
Comment (zaro0508) [inline README.md]: I think you need more than this to allow programmatic or CLI access.  If giving write access you might as well include read.  I recommend providing this policy as a reference https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_s3_rw-bucket.html

File: sam_handler/handler.py
Comment (zaro0508) [inline sam_handler/handler.py:203]: By default the SAM CLI creates builds in a `.aws-sam` folder.  why not just keep that as the default?

File: sam_handler/handler.py
Comment (zaro0508) [inline sam_handler/handler.py:187]: hmm, package does not deploy so where is the step to deploy the sam app into the AWS acccount?

File: README.md
Comment (jfalkenstein) [reply README.md]: What would be your preference?

File: README.md
Comment (jfalkenstein) [reply README.md:98]: I believe so. You could pass sam_package_args of "profile" and "region", which should take priority over the environment variables.

File: setup.py
Comment (jfalkenstein) [reply setup.py]: Oh, I made that change before you had updated the template. I can put it back.

File: README.md
Comment (jfalkenstein) [reply README.md:71]: `sam build` will compile all build dependencies for each lambda into their own assets. It doesn't actually use the local virtual environment to build outside of SAM's own dependencies.

File: README.md
Comment (jfalkenstein) [reply README.md]: I suppose that could work.

File: README.md
Comment (jfalkenstein) [reply README.md]: hmmm.... yeah, I might just link to here instead: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-permissions.html.

File: sam_handler/handler.py
Comment (jfalkenstein) [reply sam_handler/handler.py:203]: I chose not to rely on that because there are various ways to configure sam to build to a different location. Also, I'd rather be specific about where I want the file than rely on default behavior.

File: sam_handler/handler.py
Comment (jfalkenstein) [reply sam_handler/handler.py:187]: You don't need SAM to deploy. That's what Sceptre is for! `package` _does_ upload the files to S3, however.

Comment (jfalkenstein) [reply sam_handler/handler.py:187]: To clarify, `sam package` will take the built artifacts, upload them to S3, and transform the uris on the resources to point to the uploaded s3 keys rather than local file paths and outputs that transformed template. The transformed template is what Sceptre will deploy.

File: setup.py
Comment (jfalkenstein) [reply setup.py:21]: Who does that email address go to? You?

File: README.md
Comment (jfalkenstein) [reply README.md]: I kind of like "sceptre-sam-handler" better. But if you'd prefer otherwise, I'll defer to your judgment.

File: README.md
Comment (jfalkenstein) [reply README.md:98]: I'll add a note for this.

File: requirements.txt
Comment (jfalkenstein) [reply requirements.txt]: Since we're invoking sam as a subprocess, it's not required so long as SAM is accessible on the PATH. I can remove this. I'll add sam install via pypi as an "extra" that can be installed this way.

File: README.md
Comment (jfalkenstein) [reply README.md:98]: Note added (see below)

File: README.md
Comment (zaro0508) [reply README.md]: ok, lets change the repo name then.  i guess we can just infer it's a template handler from it being `sam`

File: README.md
Comment (zaro0508) [reply README.md:71]: yup, you are correct

File: setup.py
Comment (zaro0508) [reply setup.py:21]: it goes into the ether 😄

File: sam_handler/handler.py
Comment (zaro0508) [reply sam_handler/handler.py:187]: yeah,  think the thing that might help is to provide an example of a more complete sceptre template that includes the SAM template, stack_name, parameters, etc.. with this description.  I think describing what happens to the transformed (output) template and how it gets passed to sceptre to deploy the SAM app will help clarify what this handler is doing.

File: README.md
Comment (zaro0508) [inline README.md:86]: I think it would be beneficial to also deploy the transformed (output) template as well.  although i think that can be done as an improvement.

File: README.md
Comment (jfalkenstein) [reply README.md]: I changed the repo name.

File: README.md
Comment (jfalkenstein) [reply README.md:86]: Sceptre will do that for you if you've used a `template_bucket_name`. No need to bother doing that in the handler.

File: sam_handler/handler.py
Comment (jfalkenstein) [reply sam_handler/handler.py:187]: I've added both a full stack config as the example as well as a whole section on the readme for "how does this handler work"

File: README.md
Comment (zaro0508) [inline README.md]: capitalize CLI ?

File: README.md
Comment (zaro0508) [inline README.md]: work/work with

File: setup.py
Comment (zaro0508) [reply setup.py]: did you forget this?

**Review by zaro0508 — APPROVED** [review]
2 minor things but otherwise this is a sweet handler, thanks!

File: README.md
Comment (jfalkenstein) [reply README.md]: sure. I'll do that.

**Pull request merged**
